analyses/doublet-detection/exploratory-notebooks/03_compare-benchmark-results.Rmd

```diff
@@ -0,0 +1,522 @@
+        # Extract information we want to present later in a table
+        tibble::tibble(
+          "Dataset name" = dataset,
+          "Kappa" = round(confusion_result$overall["Kappa"], 3), 
+          "Balanced accuracy" = round(confusion_result$byClass["Balanced Accuracy"], 3)
+        )
+    }
+  ) |>
+  dplyr::bind_rows()
+```
+
+
+
+### Conclusions
+
+Overall, methods do not have substantial overlap with each other. 
+They each tend to detect different sets of doublets, leading to fairly small sets of consensus doublets. 
+Further, the consensus doublets called by all three methods have some, but not substantial, overlap with the ground truth. 
+
+For three out of four datasets, `scDblFinder` predicts a much larger number of doublets compared to other methods. 
+
+Based on the PCAs, it additionally looks like there are many more false negatives in the consensus prediction that `cxds` appears to capture but other methods do not.
+
+The table below summarizes performance of the "consensus caller". 
+Note that, in the [benchmarking paper these datasets were originally analyzed in](https://doi.org/10.1016/j.cels.2020.11.008), `hm-6k` was observed to be one of the "easiest" datasets to classify across methods.  
+Consistent with that observation, it has the highest `kappa` value here.
+
+```{r}
+metric_df
+```
+
+## Explore consensus of `scDblFinder` and `cxds`
+
+The above analysis suggests that `scrublet` may be more conservative than the other two methods, so including it in a consensus may be increasing the number of false positives.
+Here, we'll repeat the same consensus analyses, but considering _only_ `scDblFinder` and `cxds`.
+Since there are only two methods here, we will create two columns for this data:
+
+### Prepare data
+
+```{r}
+# modify existing consensus columns to only consider 2 methods
+doublet_df_list <- doublet_df_list |>
+  purrr::map(
+    \(dataset_df) {
+      
+      dataset_df <- dataset_df |>
+        dplyr::rowwise() |>
+        # Update column `consensus_call`
+        dplyr::mutate(
+          consensus_call = dplyr::if_else(
+            all(c(scdbl_prediction, cxds_prediction) == "doublet"),
+            "doublet", 
+            "singlet"
+          )
+        ) |>
+        # Add Update `consensus_class`
+        dplyr::mutate(
+          consensus_class = dplyr::case_when(
+            consensus_call == "doublet" ~ "doublet",
+            all(c(scdbl_prediction, cxds_prediction) == "singlet") ~ "singlet", 
+            .default = "ambiguous"
+          )
+         ) |>
+        # Add Update `confusion_call`
+        dplyr::mutate(
+          confusion_call = dplyr::case_when(
+            consensus_call == "doublet" && ground_truth == "doublet" ~ "tp",
+            consensus_call == "singlet" && ground_truth == "singlet" ~ "tn",
+            consensus_call == "doublet" && ground_truth == "singlet" ~ "fp",
+            consensus_call == "singlet" && ground_truth == "doublet" ~ "fn"
+          )
+        )
```

Comment:
I think I would pull this out to a function that you have at the top of the notebook and include an argument for `consensus_columns` or something like that where you can specify the vector of columns to use for consensus.

Reply:
Noting we discussed off github that I'm not going to port this into a function.